ycmd/completers/go/gocode_completer.py

```diff
@@ -0,0 +1,135 @@
+  then the one on the path, if not.
+
+  If the 'gocode_binary_path' is specified, use it
+  as an absolute path.
+
+  If the resolved binary exists, return the path,
+  otherwise return None.
+  '''
+  if user_options.get( 'gocode_binary_path' ):
+    # The user has explicitly specified a path.
+    if os.path.exists( user_options[ 'gocode_binary_path' ] ):
+      return user_options[ 'gocode_binary_path' ]
+    else:
+      return None
+  # Try to use the bundled binary or one on the path.
+  if os.path.exists( PATH_TO_GOCODE_BINARY ):
+    return PATH_TO_GOCODE_BINARY
+  return utils.PathToFirstExistingExecutable( [ 'gocode' ] )
+
+
+
+class GoCodeCompleter( Completer ):
+  def __init__( self, user_options ):
+    super( GoCodeCompleter, self ).__init__( user_options )
+    self._popener = utils.SafePopen # Overridden in test.
+    self._binary = FindGoCodeBinary( user_options )
+
+
+  def SupportedFiletypes( self ):
+    return GO_FILETYPES
+
+
+  def ComputeCandidatesInner( self, request_data ):
+    filename = request_data[ 'filepath' ]
+    _logger.info( "gocode completion request %s" % filename )
+    if not filename:
+      return
+
+    contents = utils.ToUtf8IfNeeded(
+        request_data['file_data'][ filename ][ 'contents' ] )
+    offset = _ComputeOffset( contents, request_data[ 'line_num' ],
+                            request_data[ 'column_num' ] )
+
+    cmd = [self._binary, '-f=json', 'autocomplete', filename, str( offset )]
+    proc = self._popener( cmd, stdin=subprocess.PIPE,
+                         stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdoutdata, stderrdata = proc.communicate( contents )
+    if proc.returncode:
+      _logger.error( COMPLETION_ERROR_MESSAGE + " code %i stderr: %s",
+                    proc.returncode, stderrdata)
+      raise RuntimeError( COMPLETION_ERROR_MESSAGE )
+
+    try:
+      resultdata = json.loads( stdoutdata )
+    except ValueError:
+      _logger.error( PARSE_ERROR_MESSAGE )
+      raise RuntimeError( PARSE_ERROR_MESSAGE )
+    if len(resultdata) != 2:
+      _logger.error( NO_COMPLETIONS_MESSAGE )
+      raise RuntimeError( NO_COMPLETIONS_MESSAGE )
+
+    return [ _ConvertCompletionData( x ) for x in resultdata[1] ]
+
+
+# Compute the byte offset in the file given the line and column.
+# TODO(ekfriis): If this is slow, consider moving this to C++ ycm_core,
+# perhaps in RequestWrap.
+def _ComputeOffset( contents, line, col ):
+  curline = 1
+  curcol = 1
+  for i, byte in enumerate( contents ):
+    if curline == line and curcol == col:
+      return i
+    curcol += 1
+    if byte == '\n':
+      curline += 1
+      curcol = 1
+  _logger.error( "GoCode completer - could not compute byte offset " +
+                "corresponding to L%i C%i", line, col )
+  return -1
+
+
+def _ConvertCompletionData( completion_data ):
+  return responses.BuildCompletionData(
+    insertion_text = completion_data[ 'name' ],
+    menu_text = completion_data[ 'name' ],
```

Comment:
As I understood it (possibly incorrectly), insertion_text is what goes into the buffer, and menu_text is what shows up on the pop-up.